master/cache.lua

```diff
@@ -0,0 +1,465 @@
+    end
+
+    return
+end
+
+
+local function fetch_and_store_state_mesos()
+    -- Fetch state JSON summary from Mesos. If successful, store to SHM cache.
+    -- Expected to run within lock context.
+    local mesosRes, err = request("leader.mesos", 5050, "/master/state-summary", false)
+
+    if err then
+        ngx.log(ngx.NOTICE, "Mesos state request failed: " .. err)
+        return
+    end
+
+    ngx.log(ngx.DEBUG, "Storing Mesos state to SHM.")
+    if not cache_data("mesosstate", mesosRes.body) then
+        ngx.log(ngx.WARN, "Storing mesos state cache failed")
+        return
+    end
+
+    ngx.update_time()
+    local time_now = ngx.now()
+    if cache_data("mesosstate_last_refresh", time_now) then
+        ngx.log(ngx.INFO, "Mesos state cache has been successfully updated")
+    end
+
+    return
+end
+
+
+local function refresh_needed(ts_name)
+    -- ts_name (str): name of the '*_last_refresh' timestamp to check
+    local cache = ngx.shared.cache
+
+    local last_fetch_time = cache:get(ts_name)
+    -- Handle the special case of first invocation.
+    if not last_fetch_time then
+        ngx.log(ngx.INFO, "Cache `".. ts_name .. "` empty. Fetching.")
+        return true
+    end
+
+    ngx.update_time()
+    local cache_age = ngx.now() - last_fetch_time
+    if cache_age > _TIMERS.CACHE_EXPIRATION then
+        ngx.log(ngx.INFO, "Cache `".. ts_name .. "` expired. Refresh.")
+        return true
+    end
+
+    ngx.log(ngx.DEBUG, "Cache `".. ts_name .. "` populated and fresh. NOOP.")
+
+    return false
+end
+
+
+local function refresh_cache(from_timer)
+    -- Refresh cache in case when it expired or has not been created yet.
+    -- Use SHM-based lock for synchronizing coroutines across worker processes.
+    --
+    -- This function can be invoked via two mechanisms:
+    --
+    --  * Via ngx.timer (in a coroutine), which is triggered
+    --    periodically in all worker processes for performing an
+    --    out-of-band cache refresh (this is the usual mode of operation).
+    --    In that case, perform cache invalidation only if no other timer
+    --    instance currently does so (abort if lock cannot immediately be
+    --    acquired).
+    --
+    --  * During HTTP request processing, when cache content is
+    --    required for answering the request but the cache was not
+    --    populated yet (i.e. usually early after nginx startup).
+    --    In that case, return from this function only after the cache
+    --    has been populated (block on lock acquisition).
+    --
+    -- Args:
+    --      from_timer: set to true if invoked from a timer
+
+    -- Acquire lock.
+    local lock
+    if from_timer then
+        ngx.log(ngx.INFO, "Executing cache refresh triggered by timer")
+        -- Fail immediately if another worker currently holds
+        -- the lock, because a single timer-based update at any
+        -- given time suffices.
+        lock = shmlock:new("shmlocks", { timeout=0 })
```

Comment:
Nice! Thanks for pointing it out!

Setting it to a constant value has some drawbacks - if for some reason, someone at some point adjusts request timeout, he/she may not be aware of `exptime` just like I did and you may not be there to catch it.

> third responds super slowly after almost timing out.

If http request lib we are using does not obey `_CONFIG.CACHE_BACKEND_REQUEST_TIMEOUT`, then unless we set it to some absurdly high value, it will never be enough.

I suggest that we set it to `3 * (_CONFIG.CACHE_BACKEND_REQUEST_TIMEOUT + 2). This way `exptime` will be always adjusted automatically. Extra 2 seconds is just an arbitrary number that is positive.

Are you OK with that?